<?xml version="1.0"?>
<robot xmlns:xacro="http://www.ros.org/wiki/xacro" name="differential_robot">

    <!-- Robot Properties -->
    <xacro:property name="base_width" value="0.360"/>  <!-- Left-right distance between wheels -->
    <xacro:property name="base_length" value="0.312"/> <!-- Front-back distance between wheels -->
    <xacro:property name="base_height" value="0.048"/>   <!-- Height of the base from ground -->
    <xacro:property name="wheel_radius" value="0.070"/> <!-- Wheel radius -->
    <xacro:property name="wheel_width" value="0.040"/>  <!-- Wheel width -->
    
    <!-- Lidar Properties -->
    <xacro:property name="lidar_height" value="0.375"/> <!-- Height from ground -->

    <!-- Colors -->
    <material name="black">
        <color rgba="0.0 0.0 0.0 1.0"/>
    </material>
    <material name="blue">
        <color rgba="0.0 0.0 0.8 1.0"/>
    </material>

    <!-- Base Link -->
    <link name="base_footprint"/>

    <joint name="base_joint" type="fixed">
        <parent link="base_footprint"/>
        <child link="base_link"/>
        <origin xyz="0 0 ${wheel_radius}" rpy="0 0 0"/>
    </joint>

    <link name="base_link">
        <visual>
            <geometry>
                <box size="${base_length} ${base_width} ${base_height}"/>
            </geometry>
            <material name="blue"/>
            <origin xyz="0 0 ${base_height/2}" rpy="0 0 0"/>
        </visual>
        <collision>                                     <!-- need for slam_toolbox and nav2-->
            <geometry>
                <box size="${base_length} ${base_width} ${base_height}"/>
            </geometry>
            <origin xyz="0 0 ${base_height/2}" rpy="0 0 0"/>
        </collision>
        <inertial>                                      <!-- don't need this tag without Gazebo -->
            <mass value="5.0"/>
            <inertia ixx="0.1" ixy="0" ixz="0" iyy="0.1" iyz="0" izz="0.1"/>
        </inertial>
    </link>

    <!-- Wheels for 2-motor differential drive -->
    <xacro:macro name="wheel" params="prefix x y">
        <link name="${prefix}_wheel">
            <visual>
                <geometry>
                    <cylinder radius="${wheel_radius}" length="${wheel_width}"/>
                </geometry>
                <material name="black"/>
                <origin xyz="0 0 0" rpy="${pi/2} 0 0"/>
            </visual>
            <collision>
                <geometry>
                    <cylinder radius="${wheel_radius}" length="${wheel_width}"/>
                </geometry>
                <origin xyz="0 0 0" rpy="${pi/2} 0 0"/>
            </collision>
            <inertial>
                <mass value="0.5"/>
                <inertia ixx="0.01" ixy="0" ixz="0" iyy="0.01" iyz="0" izz="0.01"/>
            </inertial>
        </link>

        <joint name="${prefix}_wheel_joint" type="continuous">
            <parent link="base_link"/>
            <child link="${prefix}_wheel"/>
            <origin xyz="${x} ${y} 0" rpy="0 0 0"/>
            <axis xyz="0 1 0"/>
        </joint>
    </xacro:macro>

    <!-- Create 2 wheels for differential drive -->
    <xacro:wheel prefix="left" x="0" y="${base_width/2}"/>
    <xacro:wheel prefix="right" x="0" y="-${base_width/2}"/>

    <!-- Lidar -->
    <link name="laser">
        <visual>
            <geometry>
                <cylinder radius="0.038" length="0.041"/>
            </geometry>
            <material name="black"/>
        </visual>
        <collision>
            <geometry>
                <cylinder radius="0.038" length="0.041"/>
            </geometry>
        </collision>
        <inertial>
            <mass value="0.1"/>
            <inertia ixx="0.001" ixy="0" ixz="0" iyy="0.001" iyz="0" izz="0.001"/>
        </inertial>
    </link>

    <joint name="laser_joint" type="fixed">
        <parent link="base_link"/>
        <child link="laser"/>
        <origin xyz="0 0 ${lidar_height - wheel_radius}" rpy="0 0 3.1416"/> <!-- fixed the lidar -->
    </joint>

</robot> 